src/main/cpp/src/mixed_sort_merge_join.cu

```diff
@@ -0,0 +1,825 @@
+}
+
+}  // anonymous namespace
+
+// =============================================================================
+// EQUALITY-ONLY SORT-MERGE JOIN IMPLEMENTATIONS
+// =============================================================================
+
+std::pair<rmm::device_uvector<cudf::size_type>, rmm::device_uvector<cudf::size_type>>
+sort_merge_inner_join(cudf::table_view const& left_keys,
+                      cudf::table_view const& right_keys,
+                      cudf::sorted is_left_sorted,
+                      cudf::sorted is_right_sorted,
+                      cudf::null_equality compare_nulls,
+                      rmm::cuda_stream_view stream,
+                      rmm::device_async_resource_ref mr)
+{
+  CUDF_FUNC_RANGE();
+
+  // Validate inputs
+  CUDF_EXPECTS(left_keys.num_columns() > 0, "Left keys table must have at least one column");
+  CUDF_EXPECTS(right_keys.num_columns() > 0, "Right keys table must have at least one column");
+
+  // Handle empty table cases - inner join returns empty if either table is empty
+  if (left_keys.num_rows() == 0 || right_keys.num_rows() == 0) {
+    return {rmm::device_uvector<cudf::size_type>(0, stream, mr),
+            rmm::device_uvector<cudf::size_type>(0, stream, mr)};
+  }
+
+  // Perform sort-merge inner join on equality keys
+  cudf::sort_merge_join join_obj(right_keys, is_right_sorted, compare_nulls, stream);
+  auto [left_indices, right_indices] = join_obj.inner_join(left_keys, is_left_sorted, stream, mr);
+
+  // Return the raw device_uvectors (RVO will handle the move)
+  stream.synchronize();
+  return {std::move(*left_indices), std::move(*right_indices)};
+}
+
+std::pair<rmm::device_uvector<cudf::size_type>, rmm::device_uvector<cudf::size_type>>
+sort_merge_left_join(cudf::table_view const& left_keys,
+                     cudf::table_view const& right_keys,
+                     cudf::sorted is_left_sorted,
+                     cudf::sorted is_right_sorted,
+                     cudf::null_equality compare_nulls,
+                     rmm::cuda_stream_view stream,
+                     rmm::device_async_resource_ref mr)
+{
+  CUDF_FUNC_RANGE();
+
+  // Validate inputs
+  CUDF_EXPECTS(left_keys.num_columns() > 0, "Left keys table must have at least one column");
+  CUDF_EXPECTS(right_keys.num_columns() > 0, "Right keys table must have at least one column");
+
+  // Handle empty right table case - all left rows with null right indices
+  if (right_keys.num_rows() == 0) {
+    auto left_indices  = rmm::device_uvector<cudf::size_type>(left_keys.num_rows(), stream, mr);
+    auto right_indices = rmm::device_uvector<cudf::size_type>(left_keys.num_rows(), stream, mr);
+
+    thrust::sequence(rmm::exec_policy_nosync(stream), left_indices.begin(), left_indices.end());
+    thrust::fill(rmm::exec_policy_nosync(stream),
+                 right_indices.begin(),
+                 right_indices.end(),
+                 right_keys.num_rows());
+
+    stream.synchronize();
```

Comment:
Same here, don't need stream sync.

Reply:
We do. `left_indices` and `right_indices` are in use when we exit.

Reply:
but doesn't the move make it so that we don't release the memory, so it is okay?